core/src/main/scala/kafka/server/ThrottledChannel.scala

```diff
@@ -0,0 +1,68 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.server
+
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.{Delayed, TimeUnit}
+
+import org.apache.kafka.common.utils.Time
+
+
+/**
+  * Represents a request whose response has been delayed.
+  * @param time @Time instance to use
+  * @param throttleTimeMs delay associated with this request
+  * @param channelUnmuteCallback Callback to trigger after delayTimeMs milliseconds
+  */
+class ThrottledChannel(val time: Time, val throttleTimeMs: Int, channelUnmuteCallback: () => Unit)
+  extends Delayed {
+  var endTime = time.milliseconds + throttleTimeMs
+  // Throttled channel can be unmuted after the following two events occur (order doesn't matter):
+  // (1) a response has been sent out to the client
+  // (2) throttling delay has passed
+  // This variable is incremented after each of the above events. The channel can be unmuted when the count becomes 2.
+  var count = new AtomicInteger(0)
+
+  // Used for tests
+  var channelUnmuteCallbackCalled = false
+
+  def tryUnmute() : Boolean = {
```

Comment:
It seems that the return value of `tryUnmuate` is only needed for trace level logging. Can we move the trace level logging into this method and make this method void to simplify the implementation?

Reply:
Done